flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingTracking.java

```diff
@@ -0,0 +1,171 @@
+import org.apache.flink.autoscaler.config.AutoScalerOptions;
+import org.apache.flink.autoscaler.topology.JobTopology;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/** Stores rescaling related information for the job. */
+@Experimental
+@Data
+@NoArgsConstructor
+@Builder
+public class ScalingTracking {
+
+    /** Details related to recent rescaling operations. */
+    private final TreeMap<Instant, ScalingRecord> scalingRecords = new TreeMap<>();
+
+    public void addScalingRecord(Instant startTimestamp, ScalingRecord scalingRecord) {
+        scalingRecords.put(startTimestamp, scalingRecord);
+    }
+
+    @JsonIgnore
+    public Optional<Entry<Instant, ScalingRecord>> getLatestScalingRecordEntry() {
+        if (!scalingRecords.isEmpty()) {
+            return Optional.of(scalingRecords.lastEntry());
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Sets the end time for the latest scaling record if its parallelism matches the current job
+     * parallelism.
+     *
+     * @param now The current instant to be set as the end time of the scaling record.
+     * @param jobTopology The current job topology containing details of the job's parallelism.
+     * @param scalingHistory The scaling history.
+     * @return true if the end time is successfully set, false if the end time is already set, the
+     *     latest scaling record cannot be found, or the target parallelism does not match the
+     *     actual parallelism.
+     */
+    public boolean setEndTimeIfTrackedAndParallelismMatches(
+            Instant now,
+            JobTopology jobTopology,
+            Map<JobVertexID, SortedMap<Instant, ScalingSummary>> scalingHistory) {
+        return getLatestScalingRecordEntry()
+                .map(
+                        entry -> {
+                            var value = entry.getValue();
+                            var scalingTimestamp = entry.getKey();
+                            if (value.getEndTime() == null) {
+                                var targetParallelism =
+                                        getTargetParallelismOfScaledVertices(
+                                                scalingTimestamp, scalingHistory);
+                                var actualParallelism = jobTopology.getParallelisms();
+
+                                if (targetParallelismMatchesActual(
+                                        targetParallelism, actualParallelism)) {
+                                    value.setEndTime(now);
```

Comment:
Should we maybe log this on debug? so we have an overview if we want to debug this?

Reply:
+1